newsService/src/main/java/com/kovan/app/service/NewsService.java

```diff
@@ -33,51 +39,82 @@ public class NewsService {
     private final ObjectMapper objectMapper;
     private final NewsRepositoryService service;
     private final NewsRepository newsRepository;
+    private final NewsMapper newsMapper;
+
+    LocalDate today = LocalDate.now();
+    LocalDate yesterday = today.minusDays(1);
 
     public NewsService(RestTemplate restTemplate, ObjectMapper objectMapper,
-                       NewsRepositoryService service, NewsRepository newsRepository) {
+                       NewsRepositoryService service, NewsRepository newsRepository, NewsMapper newsMapper) {
         this.restTemplate = restTemplate;
         this.objectMapper = objectMapper;
         this.service = service;
         this.newsRepository = newsRepository;
+        this.newsMapper = newsMapper;
     }
-
-    public NewsDto getTopHeadlines() {
-
-        List<NewsDto.Article> articles = new ArrayList<>();
-        int pageSize = 0;
-        int page = 1;
-        NewsDto newsDto;
-
-        do {
-            pageSize += 20;
-            String apiUrl = buildUrl(page++);
-            String response;
-            try {
-                response = restTemplate.getForObject(apiUrl, String.class);
-                newsDto = objectMapper.readValue(response, NewsDto.class);
-                articles.addAll(newsDto.getArticles());
-            } catch (JsonProcessingException e) {
-                throw new NewsRetrievalException("Failed to parse news data from API response.", e);
+    public List<NewsDto> getTopHeadlines() {
+
+        List<NewsDto.Article> articles = IntStream.iterate(1, page -> page + 1)
```

Comment:
change this logic and try to incorporate pageSize - 20 here. few refactoring needed.

Reply:
changed the logic.